### DISCUSSION TITLE
Script loading component for NextJS

### DISCUSSION THREAD

**Discussion by janicklas-ralph** [vercel/next.js#18172] [Ideas]
## Summary
We introduce a new component `next/script` to load 3P and external scripts. The component will provide a flexible and easy way to add external scripts into a NextJS application. The component will by default try to optimize the critical rendering path and ensure external scripts don't become a bottleneck to optimal page load.

Active PR - https://github.com/vercel/next.js/pull/18281

## Motivation
Third party integration is a common requirement in Next apps, where users integrate multiple script tags into the application. During our traces we’ve encountered several instances where these scripts cause performance issues and almost always block the critical rendering path.

Since all NextJS scripts required for hydration are loaded with async attributes, the user added scripts will load at a higher priority and potentially delay loading metrics such as TTI, DCL etc..

We've also run several experiments to find an ideal loading sequence for the resources in a NextJs application and the biggest problem we've seen is the unpredictability in introducing 3P scripts into the application. Users add 3P scripts at multiple points in the application and these 3P scripts can also request additional resources introducing network bottlenecks. 
This component is an attempt to bring in more context into NextJS about the external scripts so we can optimize the loading order automatically. 
The component also ensure to bring in 3P loading best practices automatically for common 3P scripts to ensure the best page load.

## Current approaches
Next provides a way to customize the scripts loaded in the document by providing a [Custom Document](https://nextjs.org/docs/advanced-features/custom-document) and [`next/head`](https://nextjs.org/docs/api-reference/next/head) to append components to <head>. Next also provides a way to dynamically import modules with [`next/dymanic`](https://nextjs.org/docs/advanced-features/dynamic-import).
Although these solve general script loading requirements, we do not have an easy way to load external 1P or 3P scripts.

Users usually add 3P scripts in the head, with next/head or modify the Document object to customize the placement of the script.
To place a script for a particular page, users generally end up adding script tags in their page components. Eg:

```
const Home = () => {
 return (
   <div>
     <script
       type="text/javascript"
       src="https://js.hsforms.net/forms/v2.js"
     />
	...
   </div>
 )
}
```

The problems with such approaches are
- Loading scripts without async/defer attribute. In cases where the script can be loaded asynchronous and isn’t.
- Potential delayed Hydration and Loading metrics. Each synchronous script tag will block the parser and delay document parsing.
- User added scripts are implicitly loaded and executed with high priority because they are included in the server-rendered HTML. [Even when it has async, a script will still cause resource contention.](https://addyosmani.com/blog/script-priorities/)

## Solution
We introduce a new `<ScriptLoader>` component that will help users handle script-loading. Any scripts added outside of the Next context with a <script> tag will be replaced with `<ScriptLoader>`.

```
import ScriptLoader from "next/script"
 
function initializeForms() {
 ...
}
 
const Home = () => {
  return (
    <div>
      <ScriptLoader src="https://js.hsforms.net/forms/v2.js"/ onLoad={initializeForms}>
	...
    </div>
  );
}
```

`<ScriptLoader>` Goals
- Streamline 3P loading in NextJs and bring in context into NextJS
- Provide different loading priorities to ensure the critical rendering path isn’t blocked by script tags
- Provide onLoad and onError callbacks
- Provide automatic optimizations/best practices recommended for 3P scripts
- NextJS will automatically optimize the placement of resources in the document to ensure the best loading sequence.
- Provide similar optimizations for inline script tags
- Switching the default approach - switches the default to be async. By using the component, we have better defaults in place.

### Loading Priorities
The component comes with 4 different loading priorities, each schedules the script at different times to support multiple use cases
- `afterInteractive` - Prioritize first party script before this script. Loads script after Next hydration. 
Example: Loading ads or analytic scripts that we want to execute as early as possible, without blocking first party scripts.
- `beforeInteractive` - Prioritize this script before first party scripts. Loads script before Hydration. 
Useful when we want to run scripts like Polyfill.io, where we want the script to execute before the first party scripts.
- `lazyOnload` - Prioritize all other resources over this script. Lazy-load the script. 
3P scripts should be ideally lazy loaded. Scripts like Social share buttons can load on button click. Below the fold ads can be loaded once the user starts scrolling. User feedback/chat scripts can load after all other primary resources have loaded.

## Additional 3P specific optimization

This is WIP. We are trying to identify common 3P script that cause page load issues and bring in automatic best practices into NextJS when we use these scripts.

**Comment by hovahyii** [comment]
May I know how to add next/script to my current Next Js project?

**Reply by Dragate** [reply]
Judging from https://github.com/vercel/next.js/tree/canary/test/integration/script-loader you need to enable its experimental flag in `next.config.js` to import `next/experimental-script`.

**Reply by hovahyii** [reply]
Can i know where can I find this next.config.js, it is under which root file?

**Reply by dsbrianwebster** [reply]
@hovahyii if you don't have one in your project you can add it to the root.

```js
module.exports = {
  experimental: { scriptLoader: true },
}
```

**Reply by dsacramone** [reply]
I am using v10.0.10-canary.7, and this isn't working. Which branch is this experiment on?

**Comment by pacocoursey** [comment]
For first-party scripts that are usually copy-pasted like Segment or GA, **could `ScriptLoader` have an `optimize` prop that would automatically compress/optimize the script contents** using Terser at build time for a smaller HTML output?

The current issue is that using `<script dangerouslySetInnerHTML>` copies the script contents exactly, including unnecessary whitespace, variables are not mangled, and no other optimizations are performed.

Ideally, this could also work for scripts that use template literals like `${GA_VERSION}`, although I'm not sure if it's possible to statically optimize those at build time since the values could change at runtime.

---

This is not applicable to third-party scripts, as those are usually optimized anyways.

**Reply by janicklas-ralph** [reply]
That's a great idea! I'll take a task to verify and add this in a future release.

**Comment by salyadav08** [comment]
I think one of the most common script used is GTM (Google Tag Manager) and unfortunately it does even come in a minified format. The solution we resort to is either go ahead with the non-minified GTM script pulled from their container or minify it in house(which is a pain to maintain). Any solution that can serve this cause?
GTM in turn downloads crazy amount of scripts. Some tack solution to not let this effect TTI would be a life saver.

**Comment by usamatekrowe** [comment]
Try nextJS Head Tag
import Head from "next/head";

**Comment by FDiskas** [comment]
How we can add script to all pages?

**Reply by dsbrianwebster** [reply]
Are you inquiring about if there is a way to do this beyond adding it in a custom _document, _app, or Layout component?

**Reply by FDiskas** [reply]
I tried to add this in _app - did not worked. Nothing was loaded. In documentation said that it should not be added in _document. So other variant would be add it to layout component?

**Comment by onurykaya** [comment]
The values ​​I wrote between the Script or ScriptLoader tags from "nexti/script" do not appear in my browser. Do you have a solution for this? The code seen below is in the <Head> tag in _documents.js.
Good work.

`                                        <ScriptLoader defer src={`https://www.googletagmanager.com/gtm/js?id=${GTM_ID}`} />

					<script defer src={`https://www.googletagmanager.com/gtag/js?id=${AW_ID}`} />

					<script defer src={`https://www.googletagmanager.com/gtag/js?id=${DC_ID}`} />

					<script defer src={`https://www.googletagmanager.com/gtag/js?id=${GTM_ID}`} />`
